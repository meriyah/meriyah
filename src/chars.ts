/**
 * A list of character constants with much more human-readable names.
 */
export const enum Chars {

    Null = 0x00,

    Backspace      = 0x08,
    Tab            = 0x09,
    LineFeed       = 0x0A,
    VerticalTab    = 0x0B,
    FormFeed       = 0x0C,
    CarriageReturn = 0x0D,

    Space       = 0x20,
    Exclamation = 0x21,
    DoubleQuote = 0x22,
    Hash        = 0x23,
    Dollar      = 0x24,
    Percent     = 0x25,
    Ampersand   = 0x26,
    SingleQuote = 0x27,
    LeftParen   = 0x28,
    RightParen  = 0x29,
    Asterisk    = 0x2A,
    Plus        = 0x2B,
    Comma       = 0x2C,
    Hyphen      = 0x2D,
    Period      = 0x2E,
    Slash       = 0x2F,

    Zero         = 0x30,
    One          = 0x31,
    Two          = 0x32,
    Three        = 0x33,
    Four         = 0x34,
    Five         = 0x35,
    Six          = 0x36,
    Seven        = 0x37,
    Eight        = 0x38,
    Nine         = 0x39,
    Colon        = 0x3A,
    Semicolon    = 0x3B,
    LessThan     = 0x3C,
    EqualSign    = 0x3D,
    GreaterThan  = 0x3E,
    QuestionMark = 0x3F,

    UpperA = 0x41,
    UpperB = 0x42,
    UpperC = 0x43,
    UpperD = 0x44,
    UpperE = 0x45,
    UpperF = 0x46,
    UpperG = 0x47,
    UpperH = 0x48,
    UpperI = 0x49,
    UpperJ = 0x4A,
    UpperK = 0x4B,
    UpperL = 0x4C,
    UpperM = 0x4D,
    UpperN = 0x4E,
    UpperO = 0x4F,

    UpperP       = 0x50,
    UpperQ       = 0x51,
    UpperR       = 0x52,
    UpperS       = 0x53,
    UpperT       = 0x54,
    UpperU       = 0x55,
    UpperV       = 0x56,
    UpperW       = 0x57,
    UpperX       = 0x58,
    UpperY       = 0x59,
    UpperZ       = 0x5A,
    LeftBracket  = 0x5B,
    Backslash    = 0x5C,
    RightBracket = 0x5D,
    Caret        = 0x5E,
    Underscore   = 0x5F,

    Backtick = 0x60,
    LowerA   = 0x61,
    LowerB   = 0x62,
    LowerC   = 0x63,
    LowerD   = 0x64,
    LowerE   = 0x65,
    LowerF   = 0x66,
    LowerG   = 0x67,
    LowerH   = 0x68,
    LowerI   = 0x69,
    LowerJ   = 0x6A,
    LowerK   = 0x6B,
    LowerL   = 0x6C,
    LowerM   = 0x6D,
    LowerN   = 0x6E,
    LowerO   = 0x6F,

    LowerP      = 0x70,
    LowerQ      = 0x71,
    LowerR      = 0x72,
    LowerS      = 0x73,
    LowerT      = 0x74,
    LowerU      = 0x75,
    LowerV      = 0x76,
    LowerW      = 0x77,
    LowerX      = 0x78,
    LowerY      = 0x79,
    LowerZ      = 0x7A,
    LeftBrace   = 0x7B,
    VerticalBar = 0x7C,
    RightBrace  = 0x7D,
    Tilde       = 0x7E,
    Delete      = 0x7F,

    NextLine         = 0x85,
    NonBreakingSpace = 0xA0,

    Ogham = 0x1680,

    EnQuad             = 0x2000,
    EmQuad             = 0x2001,
    EnSpace            = 0x2002,
    EmSpace            = 0x2003,
    ThreePerEmSpace    = 0x2004,
    FourPerEmSpace     = 0x2005,
    SixPerEmSpace      = 0x2006,
    FigureSpace        = 0x2007,
    PunctuationSpace   = 0x2008,
    ThinSpace          = 0x2009,
    HairSpace          = 0x200A,
    ZeroWidthSpace     = 0x200B,
    ZeroWidthJoiner    = 0x200C,
    ZeroWidthNonJoiner = 0x200D,

    LineSeparator      = 0x2028,
    ParagraphSeparator = 0x2029,

    NarrowNoBreakSpace = 0x202F,
    MathematicalSpace  = 0x205F,
    IdeographicSpace   = 0x3000,

    ZeroWidthNoBreakSpace = 0xFEFF,

    ByteOrderMark = 0xFFEF,

    NonBMPMax = 0x10FFFF,

    LeadSurrogateMin = 0xD800,
    LeadSurrogateMax = 0xDBFF,
    TrailSurrogateMin = 0xDC00,
    TrailSurrogateMax = 0xDFFF,
    UTF16Max = 0xFFFF,
    NonBMPMin = 0x10000,
}
