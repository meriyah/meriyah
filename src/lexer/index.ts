export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
