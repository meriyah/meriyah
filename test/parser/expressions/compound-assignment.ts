import * as t from 'node:assert/strict';
import { describe, it } from 'vitest';
import { parseSource } from '../../../src/parser';
import { pass } from '../../test-utils';

describe('Expressions - Compound assignment', () => {
  for (const arg of [
    '[a >>>= a];',
    '[a >>>= a += a];',
    '[a >>>= (a += a)];',
    '[a >>>= (a += (a))];',
    '[a >>>= a += {a}];',
    '[a >>>= a += {a}];',
    '[a >>>= a += a];',
    '([a += a] );',
    '([...a += a] );',
    '[a >>>= (a)];',
    '([...a += a += a += (a) >>>= 2]);',
    '[...a %= (a)];',
    'obj.prop >>= 20;',
    'a |= 2;',
    'obj.prop &= 20;',
    'obj.len ^= 10;',
    'var z = (x += 1);',
    'var z = (x <<= 1);',
    'x -= 1 ',
    'y1 = (y %= 2);',
    'y1 === -1',
    'x *= "1";',
    'x /= null;',
    'x >>>= true;',
    'if (scope.x !== 2) {}',
    'x /= y',
    'base[prop] /= expr();',
    'x	>>=	1, 0',
    'x	*=	-1',
    '({a: a *= -1})',
    '([a *= -1])',
    '([(a *= -1)])',
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`);
      });
    });
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { webcompat: true });
      });
    });
  }

  for (const arg of [
    '({a *= -1})',
    '({a} *= -1)',
    '({a}) *=	-1',
    '({a} += a);',
    '([a] += a);',
    '({a} += {a});',
    '[a >>>= {a} += {a}];',
    '[1 >>>= a];',
    '[a >>>= a] += 1;',
    '[a >>>= a] += a;',
    '({a: (b = 0)} = {})',
    '([(a = b)] = []',
    '({a: b += 0} = {})',
    '[a += b] = []',
    '0.toString',
    '0.toString',
    '1 >>>= 1;',
    '1 -= 1;',
    '1 &= 1;',
    '1 |= 1;',
    '1 = 1;',
  ]) {
    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`, { next: true, sourceType: 'module' });
      });
    });
    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`, { next: true });
      });
    });
    it(`${arg}`, () => {
      t.throws(() => {
        parseSource(`${arg}`, { webcompat: true });
      });
    });
  }

  pass('Expressions - Compound assignment (pass)', [
    { code: 'base[prop()] /= expr();', options: { ranges: true } },
    { code: 'null && (x += null)', options: { ranges: true } },
    { code: 'y1 = (y %= 2);', options: { ranges: true } },
    'y1 = (y <<= 1);',
    { code: 'x ^= new String("1");', options: { ranges: true } },
    'x *= "1";',
    { code: 'obj.prop >>= 20;', options: { ranges: true } },
    'arguments &= 20;',
    { code: 'var z = (x *= -1);', options: { ranges: true } },
    'var z = (x %= y);',
    'x *= undefined;',
    'x -= 1;',
    { code: '(new foo).bar()', options: { ranges: true } },
    { code: 'a.b.c(2020)', options: { ranges: true } },
    { code: 'a(0).b(14, 3, 77).c', options: { ranges: true } },
    { code: 'x >>= 1;', options: { ranges: true } },
    'var x1 = (x <<= 1);',
    'x |= 1;',
    'x /= true;',
    'obj.len &= 10;',
    'var x = 4;',
    '(x + y) >= z',
    '(x + y) <= z',
    'x *= undefined;',
    'x *= null;',
    'x |= "1";',
    'z = (x %= y);',
    'x += "1";',
    'x >>= true;',

    'x |= true',

    'x |= "1"',
    'x |= 1',
    'x += true',
  ]);
});
