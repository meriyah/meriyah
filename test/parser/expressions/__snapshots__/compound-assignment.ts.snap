// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > (new foo).bar() 1`] = `
{
  "body": [
    {
      "end": 15,
      "expression": {
        "arguments": [],
        "callee": {
          "computed": false,
          "end": 13,
          "object": {
            "arguments": [],
            "callee": {
              "end": 8,
              "name": "foo",
              "range": [
                5,
                8,
              ],
              "start": 5,
              "type": "Identifier",
            },
            "end": 8,
            "range": [
              1,
              8,
            ],
            "start": 1,
            "type": "NewExpression",
          },
          "optional": false,
          "property": {
            "end": 13,
            "name": "bar",
            "range": [
              10,
              13,
            ],
            "start": 10,
            "type": "Identifier",
          },
          "range": [
            0,
            13,
          ],
          "start": 0,
          "type": "MemberExpression",
        },
        "end": 15,
        "optional": false,
        "range": [
          0,
          15,
        ],
        "start": 0,
        "type": "CallExpression",
      },
      "range": [
        0,
        15,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 15,
  "range": [
    0,
    15,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > (x + y) <= z 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "left": {
            "name": "x",
            "type": "Identifier",
          },
          "operator": "+",
          "right": {
            "name": "y",
            "type": "Identifier",
          },
          "type": "BinaryExpression",
        },
        "operator": "<=",
        "right": {
          "name": "z",
          "type": "Identifier",
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > (x + y) >= z 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "left": {
            "name": "x",
            "type": "Identifier",
          },
          "operator": "+",
          "right": {
            "name": "y",
            "type": "Identifier",
          },
          "type": "BinaryExpression",
        },
        "operator": ">=",
        "right": {
          "name": "z",
          "type": "Identifier",
        },
        "type": "BinaryExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > a(0).b(14, 3, 77).c 1`] = `
{
  "body": [
    {
      "end": 19,
      "expression": {
        "computed": false,
        "end": 19,
        "object": {
          "arguments": [
            {
              "end": 9,
              "range": [
                7,
                9,
              ],
              "start": 7,
              "type": "Literal",
              "value": 14,
            },
            {
              "end": 12,
              "range": [
                11,
                12,
              ],
              "start": 11,
              "type": "Literal",
              "value": 3,
            },
            {
              "end": 16,
              "range": [
                14,
                16,
              ],
              "start": 14,
              "type": "Literal",
              "value": 77,
            },
          ],
          "callee": {
            "computed": false,
            "end": 6,
            "object": {
              "arguments": [
                {
                  "end": 3,
                  "range": [
                    2,
                    3,
                  ],
                  "start": 2,
                  "type": "Literal",
                  "value": 0,
                },
              ],
              "callee": {
                "end": 1,
                "name": "a",
                "range": [
                  0,
                  1,
                ],
                "start": 0,
                "type": "Identifier",
              },
              "end": 4,
              "optional": false,
              "range": [
                0,
                4,
              ],
              "start": 0,
              "type": "CallExpression",
            },
            "optional": false,
            "property": {
              "end": 6,
              "name": "b",
              "range": [
                5,
                6,
              ],
              "start": 5,
              "type": "Identifier",
            },
            "range": [
              0,
              6,
            ],
            "start": 0,
            "type": "MemberExpression",
          },
          "end": 17,
          "optional": false,
          "range": [
            0,
            17,
          ],
          "start": 0,
          "type": "CallExpression",
        },
        "optional": false,
        "property": {
          "end": 19,
          "name": "c",
          "range": [
            18,
            19,
          ],
          "start": 18,
          "type": "Identifier",
        },
        "range": [
          0,
          19,
        ],
        "start": 0,
        "type": "MemberExpression",
      },
      "range": [
        0,
        19,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 19,
  "range": [
    0,
    19,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > a.b.c(2020) 1`] = `
{
  "body": [
    {
      "end": 11,
      "expression": {
        "arguments": [
          {
            "end": 10,
            "range": [
              6,
              10,
            ],
            "start": 6,
            "type": "Literal",
            "value": 2020,
          },
        ],
        "callee": {
          "computed": false,
          "end": 5,
          "object": {
            "computed": false,
            "end": 3,
            "object": {
              "end": 1,
              "name": "a",
              "range": [
                0,
                1,
              ],
              "start": 0,
              "type": "Identifier",
            },
            "optional": false,
            "property": {
              "end": 3,
              "name": "b",
              "range": [
                2,
                3,
              ],
              "start": 2,
              "type": "Identifier",
            },
            "range": [
              0,
              3,
            ],
            "start": 0,
            "type": "MemberExpression",
          },
          "optional": false,
          "property": {
            "end": 5,
            "name": "c",
            "range": [
              4,
              5,
            ],
            "start": 4,
            "type": "Identifier",
          },
          "range": [
            0,
            5,
          ],
          "start": 0,
          "type": "MemberExpression",
        },
        "end": 11,
        "optional": false,
        "range": [
          0,
          11,
        ],
        "start": 0,
        "type": "CallExpression",
      },
      "range": [
        0,
        11,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 11,
  "range": [
    0,
    11,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > arguments &= 20; 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "arguments",
          "type": "Identifier",
        },
        "operator": "&=",
        "right": {
          "type": "Literal",
          "value": 20,
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > base[prop()] /= expr(); 1`] = `
{
  "body": [
    {
      "end": 23,
      "expression": {
        "end": 22,
        "left": {
          "computed": true,
          "end": 12,
          "object": {
            "end": 4,
            "name": "base",
            "range": [
              0,
              4,
            ],
            "start": 0,
            "type": "Identifier",
          },
          "optional": false,
          "property": {
            "arguments": [],
            "callee": {
              "end": 9,
              "name": "prop",
              "range": [
                5,
                9,
              ],
              "start": 5,
              "type": "Identifier",
            },
            "end": 11,
            "optional": false,
            "range": [
              5,
              11,
            ],
            "start": 5,
            "type": "CallExpression",
          },
          "range": [
            0,
            12,
          ],
          "start": 0,
          "type": "MemberExpression",
        },
        "operator": "/=",
        "range": [
          0,
          22,
        ],
        "right": {
          "arguments": [],
          "callee": {
            "end": 20,
            "name": "expr",
            "range": [
              16,
              20,
            ],
            "start": 16,
            "type": "Identifier",
          },
          "end": 22,
          "optional": false,
          "range": [
            16,
            22,
          ],
          "start": 16,
          "type": "CallExpression",
        },
        "start": 0,
        "type": "AssignmentExpression",
      },
      "range": [
        0,
        23,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 23,
  "range": [
    0,
    23,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > null && (x += null) 1`] = `
{
  "body": [
    {
      "end": 19,
      "expression": {
        "end": 19,
        "left": {
          "end": 4,
          "range": [
            0,
            4,
          ],
          "start": 0,
          "type": "Literal",
          "value": null,
        },
        "operator": "&&",
        "range": [
          0,
          19,
        ],
        "right": {
          "end": 18,
          "left": {
            "end": 10,
            "name": "x",
            "range": [
              9,
              10,
            ],
            "start": 9,
            "type": "Identifier",
          },
          "operator": "+=",
          "range": [
            9,
            18,
          ],
          "right": {
            "end": 18,
            "range": [
              14,
              18,
            ],
            "start": 14,
            "type": "Literal",
            "value": null,
          },
          "start": 9,
          "type": "AssignmentExpression",
        },
        "start": 0,
        "type": "LogicalExpression",
      },
      "range": [
        0,
        19,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 19,
  "range": [
    0,
    19,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > obj.len &= 10; 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "computed": false,
          "object": {
            "name": "obj",
            "type": "Identifier",
          },
          "optional": false,
          "property": {
            "name": "len",
            "type": "Identifier",
          },
          "type": "MemberExpression",
        },
        "operator": "&=",
        "right": {
          "type": "Literal",
          "value": 10,
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > obj.prop >>= 20; 1`] = `
{
  "body": [
    {
      "end": 16,
      "expression": {
        "end": 15,
        "left": {
          "computed": false,
          "end": 8,
          "object": {
            "end": 3,
            "name": "obj",
            "range": [
              0,
              3,
            ],
            "start": 0,
            "type": "Identifier",
          },
          "optional": false,
          "property": {
            "end": 8,
            "name": "prop",
            "range": [
              4,
              8,
            ],
            "start": 4,
            "type": "Identifier",
          },
          "range": [
            0,
            8,
          ],
          "start": 0,
          "type": "MemberExpression",
        },
        "operator": ">>=",
        "range": [
          0,
          15,
        ],
        "right": {
          "end": 15,
          "range": [
            13,
            15,
          ],
          "start": 13,
          "type": "Literal",
          "value": 20,
        },
        "start": 0,
        "type": "AssignmentExpression",
      },
      "range": [
        0,
        16,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 16,
  "range": [
    0,
    16,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > var x = 4; 1`] = `
{
  "body": [
    {
      "declarations": [
        {
          "id": {
            "name": "x",
            "type": "Identifier",
          },
          "init": {
            "type": "Literal",
            "value": 4,
          },
          "type": "VariableDeclarator",
        },
      ],
      "kind": "var",
      "type": "VariableDeclaration",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > var x1 = (x <<= 1); 1`] = `
{
  "body": [
    {
      "declarations": [
        {
          "id": {
            "name": "x1",
            "type": "Identifier",
          },
          "init": {
            "left": {
              "name": "x",
              "type": "Identifier",
            },
            "operator": "<<=",
            "right": {
              "type": "Literal",
              "value": 1,
            },
            "type": "AssignmentExpression",
          },
          "type": "VariableDeclarator",
        },
      ],
      "kind": "var",
      "type": "VariableDeclaration",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > var z = (x %= y); 1`] = `
{
  "body": [
    {
      "declarations": [
        {
          "id": {
            "name": "z",
            "type": "Identifier",
          },
          "init": {
            "left": {
              "name": "x",
              "type": "Identifier",
            },
            "operator": "%=",
            "right": {
              "name": "y",
              "type": "Identifier",
            },
            "type": "AssignmentExpression",
          },
          "type": "VariableDeclarator",
        },
      ],
      "kind": "var",
      "type": "VariableDeclaration",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > var z = (x *= -1); 1`] = `
{
  "body": [
    {
      "declarations": [
        {
          "end": 17,
          "id": {
            "end": 5,
            "name": "z",
            "range": [
              4,
              5,
            ],
            "start": 4,
            "type": "Identifier",
          },
          "init": {
            "end": 16,
            "left": {
              "end": 10,
              "name": "x",
              "range": [
                9,
                10,
              ],
              "start": 9,
              "type": "Identifier",
            },
            "operator": "*=",
            "range": [
              9,
              16,
            ],
            "right": {
              "argument": {
                "end": 16,
                "range": [
                  15,
                  16,
                ],
                "start": 15,
                "type": "Literal",
                "value": 1,
              },
              "end": 16,
              "operator": "-",
              "prefix": true,
              "range": [
                14,
                16,
              ],
              "start": 14,
              "type": "UnaryExpression",
            },
            "start": 9,
            "type": "AssignmentExpression",
          },
          "range": [
            4,
            17,
          ],
          "start": 4,
          "type": "VariableDeclarator",
        },
      ],
      "end": 18,
      "kind": "var",
      "range": [
        0,
        18,
      ],
      "start": 0,
      "type": "VariableDeclaration",
    },
  ],
  "end": 18,
  "range": [
    0,
    18,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x *= "1"; 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "*=",
        "right": {
          "type": "Literal",
          "value": "1",
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x *= null; 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "*=",
        "right": {
          "type": "Literal",
          "value": null,
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x *= undefined; 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "*=",
        "right": {
          "name": "undefined",
          "type": "Identifier",
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x *= undefined; 2`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "*=",
        "right": {
          "name": "undefined",
          "type": "Identifier",
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x += "1"; 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "+=",
        "right": {
          "type": "Literal",
          "value": "1",
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x += true 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "+=",
        "right": {
          "type": "Literal",
          "value": true,
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x /= true; 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "/=",
        "right": {
          "type": "Literal",
          "value": true,
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x >>= 1; 1`] = `
{
  "body": [
    {
      "end": 8,
      "expression": {
        "end": 7,
        "left": {
          "end": 1,
          "name": "x",
          "range": [
            0,
            1,
          ],
          "start": 0,
          "type": "Identifier",
        },
        "operator": ">>=",
        "range": [
          0,
          7,
        ],
        "right": {
          "end": 7,
          "range": [
            6,
            7,
          ],
          "start": 6,
          "type": "Literal",
          "value": 1,
        },
        "start": 0,
        "type": "AssignmentExpression",
      },
      "range": [
        0,
        8,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 8,
  "range": [
    0,
    8,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x >>= true; 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": ">>=",
        "right": {
          "type": "Literal",
          "value": true,
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x ^= new String("1"); 1`] = `
{
  "body": [
    {
      "end": 21,
      "expression": {
        "end": 20,
        "left": {
          "end": 1,
          "name": "x",
          "range": [
            0,
            1,
          ],
          "start": 0,
          "type": "Identifier",
        },
        "operator": "^=",
        "range": [
          0,
          20,
        ],
        "right": {
          "arguments": [
            {
              "end": 19,
              "range": [
                16,
                19,
              ],
              "start": 16,
              "type": "Literal",
              "value": "1",
            },
          ],
          "callee": {
            "end": 15,
            "name": "String",
            "range": [
              9,
              15,
            ],
            "start": 9,
            "type": "Identifier",
          },
          "end": 20,
          "range": [
            5,
            20,
          ],
          "start": 5,
          "type": "NewExpression",
        },
        "start": 0,
        "type": "AssignmentExpression",
      },
      "range": [
        0,
        21,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 21,
  "range": [
    0,
    21,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x |= "1" 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "|=",
        "right": {
          "type": "Literal",
          "value": "1",
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x |= "1"; 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "|=",
        "right": {
          "type": "Literal",
          "value": "1",
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x |= 1 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "|=",
        "right": {
          "type": "Literal",
          "value": 1,
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x |= 1; 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "|=",
        "right": {
          "type": "Literal",
          "value": 1,
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x |= true 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "|=",
        "right": {
          "type": "Literal",
          "value": true,
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > x -= 1; 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "-=",
        "right": {
          "type": "Literal",
          "value": 1,
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > y1 = (y %= 2); 1`] = `
{
  "body": [
    {
      "end": 14,
      "expression": {
        "end": 13,
        "left": {
          "end": 2,
          "name": "y1",
          "range": [
            0,
            2,
          ],
          "start": 0,
          "type": "Identifier",
        },
        "operator": "=",
        "range": [
          0,
          13,
        ],
        "right": {
          "end": 12,
          "left": {
            "end": 7,
            "name": "y",
            "range": [
              6,
              7,
            ],
            "start": 6,
            "type": "Identifier",
          },
          "operator": "%=",
          "range": [
            6,
            12,
          ],
          "right": {
            "end": 12,
            "range": [
              11,
              12,
            ],
            "start": 11,
            "type": "Literal",
            "value": 2,
          },
          "start": 6,
          "type": "AssignmentExpression",
        },
        "start": 0,
        "type": "AssignmentExpression",
      },
      "range": [
        0,
        14,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  ],
  "end": 14,
  "range": [
    0,
    14,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > y1 = (y <<= 1); 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "y1",
          "type": "Identifier",
        },
        "operator": "=",
        "right": {
          "left": {
            "name": "y",
            "type": "Identifier",
          },
          "operator": "<<=",
          "right": {
            "type": "Literal",
            "value": 1,
          },
          "type": "AssignmentExpression",
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Expressions - Compound assignment > Expressions - Compound assignment (pass) > z = (x %= y); 1`] = `
{
  "body": [
    {
      "expression": {
        "left": {
          "name": "z",
          "type": "Identifier",
        },
        "operator": "=",
        "right": {
          "left": {
            "name": "x",
            "type": "Identifier",
          },
          "operator": "%=",
          "right": {
            "name": "y",
            "type": "Identifier",
          },
          "type": "AssignmentExpression",
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;
